### PR TITLE
Blog post: Using Qute with templates from a database

### PR DESCRIPTION
**Pull request by gwenneg**

### CODE REVIEW

**Comment by gwenneg** [comment]
@mkouba Could you please have a look at this blog post?

**Comment by gwenneg** [comment]
The code snippet about `DbTemplateLocator` contained something wrong, I fixed it.

**Comment by mkouba** [comment]
> This won’t work with the Qute engine instance provided by Quarkus though.

It would be useful to mention why it doesn't work...

FTR are you aware of the `@io.quarkus.qute.Locate` and [Template Locator Registration](https://quarkus.io/version/main/guides/qute-reference#template-locator-registration)? This is available in 2.10.

**Comment by gwenneg** [comment]
> > This won’t work with the Qute engine instance provided by Quarkus though.
> 
> It would be useful to mention why it doesn't work...
> 
> FTR are you aware of the `@io.quarkus.qute.Locate` and [Template Locator Registration](https://quarkus.io/version/main/guides/qute-reference#template-locator-registration)? This is available in 2.10.

No I wasn't aware of this feature, thanks!

Does it mean we could use DB templates with the Qute engine instance provided by Quarkus 2.10? If so, would features like `@TemplateExtension` also work with the DB templates?

**Comment by gwenneg** [comment]
> It would be useful to mention why it doesn't work...

I'm not sure how to explain that. 😃 The Qute engine provided by Quarkus cannot be configured using the `Engine.builder()` API, that's all I know.

**Comment by mkouba** [comment]
> Does it mean we could use DB templates with the Qute engine instance provided by Quarkus 2.10? If so, would features like `@TemplateExtension` also work with the DB templates?

Yes, this should work.

> I'm not sure how to explain that. smiley The Qute engine provided by Quarkus cannot be configured using the Engine.builder() API, that's all I know.

Well, you can use a CDI observer to modify the default `EngineBuilder`, see https://quarkus.io/guides/qute-reference#quarkus_integration.

**Comment by gwenneg** [comment]
That's another interesting thing I wasn't aware of. It looks like I have a few tests to do before updating the blog post.

Do you think it's interesting to keep our old way of registering the template locator while mentioning the new feature from 2.10? Or should the blog post only talk about 2.10?

**Comment by mkouba** [comment]
> Do you think it's interesting to keep in the post our old way of registering the template locator while mentioning the new feature from 2.10? Or should the blog post only talk about 2.10?

I think that it would be great to have both. First of all not all users can upgrade to the latest version and also it shows that quarkus is evolving ;-)

**Comment by gwenneg** [comment]
I gave the CDI observer a try and it doesn't work for us (at least in tests) because Quarkus tries to load a template from the DB before we get a chance to open the Hibernate stateless session required by our template locator. Our real code is very different from the simple version I put in the blog post. 😄 This observer thing still needs to be mentioned so I'll update the post to include that.

I'll mention the 2.10 feature as well.

**Comment by mkouba** [comment]
> ...because Quarkus tries to load a template from the DB before we get a chance to open the Hibernate stateless session required by our template locator.

When exactly is quarkus trying to load the template? Only the discovered templates should be loaded eagerly in the `EngineProducer`...

**Comment by gwenneg** [comment]
Our build failed with the following cause at application startup before any test could be executed:
```
Caused by: java.lang.IllegalStateException: Current stateless session not found. Did you use StatelessSessionFactory#withSession before calling this method?
	at com.redhat.cloud.notifications.db.StatelessSessionFactory.getCurrentSession(StatelessSessionFactory.java:62)
	at com.redhat.cloud.notifications.db.StatelessSessionFactory_ClientProxy.getCurrentSession(Unknown Source)
	at com.redhat.cloud.notifications.templates.DbTemplateLocator.locate(DbTemplateLocator.java:26)
	at com.redhat.cloud.notifications.templates.DbTemplateLocator_ClientProxy.locate(Unknown Source)
	at io.quarkus.qute.EngineImpl.load(EngineImpl.java:160)
	at java.base/java.util.concurrent.ConcurrentHashMap.computeIfAbsent(ConcurrentHashMap.java:1705)
	at io.quarkus.qute.EngineImpl.getTemplate(EngineImpl.java:121)
	at io.quarkus.qute.runtime.EngineProducer.<init>(EngineProducer.java:236)
	at io.quarkus.qute.runtime.EngineProducer_Bean.create(Unknown Source)
	at io.quarkus.qute.runtime.EngineProducer_Bean.create(Unknown Source)
	at io.quarkus.arc.impl.AbstractSharedContext.createInstanceHandle(AbstractSharedContext.java:111)
	at io.quarkus.arc.impl.AbstractSharedContext$1.get(AbstractSharedContext.java:35)
	at io.quarkus.arc.impl.AbstractSharedContext$1.get(AbstractSharedContext.java:32)
	at io.quarkus.arc.impl.LazyValue.get(LazyValue.java:26)
	at io.quarkus.arc.impl.ComputingCache.computeIfAbsent(ComputingCache.java:69)
	at io.quarkus.arc.impl.AbstractSharedContext.get(AbstractSharedContext.java:32)
	at io.quarkus.qute.runtime.EngineProducer_Bean.get(Unknown Source)
	at io.quarkus.qute.runtime.EngineProducer_Bean.get(Unknown Source)
	at io.quarkus.arc.impl.ArcContainerImpl.beanInstanceHandle(ArcContainerImpl.java:467)
	at io.quarkus.arc.impl.ArcContainerImpl.beanInstanceHandle(ArcContainerImpl.java:480)
	at io.quarkus.arc.impl.ArcContainerImpl.instance(ArcContainerImpl.java:292)
	at io.quarkus.qute.runtime.EngineProducer_Observer_Synthetic_5ca7449e3ab1631e7e8ed910338051c00d29a79b.notify(Unknown Source)
	at io.quarkus.arc.impl.EventImpl$Notifier.notifyObservers(EventImpl.java:323)
	at io.quarkus.arc.impl.EventImpl$Notifier.notify(EventImpl.java:305)
	at io.quarkus.arc.impl.EventImpl.fire(EventImpl.java:73)
	at io.quarkus.arc.runtime.ArcRecorder.fireLifecycleEvent(ArcRecorder.java:130)
	at io.quarkus.arc.runtime.ArcRecorder.handleLifecycleEvents(ArcRecorder.java:99)
	at io.quarkus.deployment.steps.LifecycleEventsBuildStep$startupEvent1144526294.deploy_0(Unknown Source)
	at io.quarkus.deployment.steps.LifecycleEventsBuildStep$startupEvent1144526294.deploy(Unknown Source)
	... 52 more
```

**Comment by gwenneg** [comment]
I updated the post with details about the template locator registration alternatives.

**Comment by mkouba** [comment]
> Our build failed with the following cause at application startup before any test could be executed:

That's probably because your template locator is also used for regular templates. Two solutions come to my mind:
1. Implement `TemplateLocator#getPriority()` in your locator and return a value lower than `io.quarkus.qute.WithPriority.DEFAULT_PRIORITY` (`1`). This way the default locator will be used for the discovered templates duing app init.
2. Use a specific prefix for the templates stored in the DB so that you can return `Optional.empty()` without a DB hit.

**Comment by gwenneg** [comment]
Overriding the priority and setting it to `0` worked 👍 I'll update the post to remove the part about the custom Qute engine instance.

**Comment by gwenneg** [comment]
@mkouba I'm done making big changes. I also reviewed in details the entire post and fixed all typos mistakes I found.

If the current version is OK with you from a Qute perspective, I will probably add a few diagrams/drawings later today.

**Comment by mkouba** [comment]
It looks great!

I wonder if it would make sense to improve the API and simplify the removal from the cache somehow... I'm not sure how though ;-).

**Comment by gwenneg** [comment]
I tried to include an image in the post. I hope I did that right :smile:

This PR is ready for a "final" review.

File: _posts/2022-06-20-qute-templates-from-db.adoc [hunk 1]
```diff
@@ -0,0 +1,282 @@
...
+
+import javax.enterprise.event.Observes;
+import javax.inject.Inject;
+
+public class QuteEngineConfiguration {
+
+    @Inject
+    DbTemplateLocator dbTemplateLocator;
+
+    void configureEngine(@Observes EngineBuilder builder) {
```
Comment (mkouba) [inline _posts/2022-06-20-qute-templates-from-db.adoc:133]: In fact, you could try to add this observer method directly to the `DbTemplateLocator`, i.e.:
```java
void configureEngine(@Observes EngineBuilder builder) {
   builder.addLocator(this);
}
```

Comment (gwenneg) [reply _posts/2022-06-20-qute-templates-from-db.adoc]: I updated the post with that suggestion, thanks!

**Comment by gwenneg** [comment]
I'll squash the commits before the merge.

**Review by mkouba — APPROVED** [review]
The content looks great. Thanks!

CC @gsmet

**Comment by gwenneg** [comment]
Commits squashed.

Thanks for the review @mkouba!

**Comment by gwenneg** [comment]
Date changed to today.

**Review by gsmet — APPROVED** [review]

File: _posts/2022-06-21-qute-templates-from-db.adoc [hunk 2]
```diff
@@ -0,0 +1,267 @@
...
+public class TemplateRepository {
+
+    @Inject
+    EntityManager entityManager;
+
+    @Transactional
+    public void deleteTemplate(String name) {
+        long count = entityManager.createQuery("SELECT COUNT(*) FROM DbTemplate WHERE name != :name AND content LIKE :include", Long.class)
+                .setParameter("name", name)
+                .setParameter("include", "%{#include " + name + "%")
```
Comment (gastaldi) [inline _posts/2022-06-21-qute-templates-from-db.adoc:246]: A little late to the game, but isn't this line missing a `}`?

Comment (gastaldi) [reply _posts/2022-06-21-qute-templates-from-db.adoc:246]: Ah nevermind, it's a `LIKE` statement, I was looking at the parameter and forgot about the SQL above :)

**Pull request merged**